src/nupic/utils/StlIo.hpp

```diff
@@ -32,126 +32,67 @@
-  pf(out_stream);
-  return out_stream;
-}
-
-
-struct abbr {
-  int n;
-  inline abbr(int _n) : n(_n) {}
-};
-
-template <typename CharT, typename Traits>
-inline std::basic_ostream<CharT, Traits> &
-operator<<(std::basic_ostream<CharT, Traits> &out_stream, abbr s) {
-  io_control.abbr = s.n;
-  return out_stream;
-}
-
-struct debug {
-  int n;
-  inline debug(int _n = -1) : n(_n) {}
-};
-
-template <typename CharT, typename Traits>
-inline std::basic_ostream<CharT, Traits> &
-operator<<(std::basic_ostream<CharT, Traits> &out_stream, debug d) {
-  io_control.abbr = d.n;
-  io_control.output_n_elts = false;
-  io_control.pair_sep = ",";
-  io_control.pair_paren = true;
-  return out_stream;
-}
-
 //--------------------------------------------------------------------------------
 // BINARY PERSISTENCE
 //--------------------------------------------------------------------------------
-template <typename It>
-inline void binary_save(std::ostream &out_stream, It begin, It end) {
-  typedef typename std::iterator_traits<It>::value_type value_type;
-  size_t size = (size_t)(end - begin);
-  if (size > 0) {
+template <typename Container> //TODO Container must provide begin,end, check for that
+inline void binary_save(std::ostream &out_stream, const Container &objFrom) {
+  const auto begin = std::begin(objFrom); //iterator.begin()
+  const auto end   = std::end(objFrom);
+  using std::cout;
+  using std::endl;
+
+  const size_t size = (size_t) (end - begin);
+  out_stream << "BIN" << " ";
+  out_stream << size << " ";
+  cout << "XXX1 " << size << endl;
+
+  if(size > 0) {
+    cout << "DEBUG vector[" << *begin << "]  ";
     char *ptr = (char *)&*begin;
+    typedef typename Container::value_type value_type;
+    cout << "xxx2 " <<sizeof(value_type) << endl;
     out_stream.write(ptr, (std::streamsize)size * sizeof(value_type));
   }
+  out_stream << " ~BIN " << endl;
 }
 
 //--------------------------------------------------------------------------------
-template <typename It>
-inline void binary_load(std::istream &in_stream, It begin, It end) {
-  typedef typename std::iterator_traits<It>::value_type value_type;
-  size_t size = (size_t)(end - begin);
-  if (size > 0) {
-    char *ptr = (char *)&*begin;
-    in_stream.read(ptr, (std::streamsize)size * sizeof(value_type));
+template <typename Container>
+inline void binary_load(std::istream &in_stream, Container &objTo) {
+  std::string label;
+  in_stream >> label;
+  NTA_CHECK(label == "BIN") << "Stream mismatched.";
+  size_t sz;
+  in_stream >> sz;
+
+  using std::cout;
+using std::endl;
+  cout << "YYY1 " << sz << endl;
+
+  if(sz > 0) {    
+    objTo.resize(sz);
+    const auto begin = std::begin(objTo);
+    typedef typename Container::value_type value_type;
+    cout << "YYY2 " << sizeof(value_type) << endl;
+    char* raw = (char*)&*begin;
+    cout << "YYY 3 " << in_stream.tellg() << endl;
+    in_stream.read(raw, (std::streamsize)sz * sizeof(value_type));
```

Comment:
Do we need this function?  I think it is trying to load an arbitrary object and that would probably not work if the object contains any pointers.   If this is trying to work with a vector, it can get the destination pointer by getting the address of the first element.  That works because the vector's buffer is contiguous.  But that does not work with std::set or std::map etc. because each element's pair is allocated independently.

Reply:
> But that does not work with std::set or std::map etc. because each element's pair is allocated independently.

perfect, so I'll remove the binary_save, binary_load (as only works for vectors!), and I'll make work the StlIo::vector_save() which is "human readable, text" format.